zerver/tests/test_signup.py

```diff
@@ -1065,7 +1067,7 @@ def try_invite() -> HttpResponse:
         #       the large number of queries), so I just
         #       use an approximate equality check.
         actual_count = len(queries)
-        expected_count = 251
+        expected_count = 313
```

Comment:
Here we send 30 invites, so according to 2 query per invite (one for invitation email and other for invitation reminder email), there should be increase of 60 queries, not sure of the reason of 2 additional queries.

Reply:
This is a pretty big increase!  I think we can fix this with caching, or an optional parameter to the function to send confirmation emails (to avoid fetching the same parameter N times in a loop, which is what we're doing here).

I'm not sure why this number is 250+ in the first place; but I guess there's a TODO comment.  It would be a good use of time to debug that in a separate PR; for this one, I think it's probably at least worth avoiding another another 60.

Reply:
```diff
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index c4e66e66d2..b01177e373 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -6624,7 +6624,7 @@ def filter_presence_idle_user_ids(user_ids: Set[int]) -> List[int]:
     return sorted(idle_user_ids)
 
 
-def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile) -> str:
+def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile, language: str) -> str:
     """
     Send the confirmation/welcome e-mail to an invited user.
     """
@@ -6635,12 +6635,11 @@ def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfi
         "activate_url": activation_url,
         "referrer_realm_name": referrer.realm.name,
     }
-    realm_user_default = RealmUserDefault.objects.get(realm=referrer.realm)
     send_email(
         "zerver/emails/invitation",
         to_emails=[invitee.email],
         from_address=FromAddress.tokenized_no_reply_address(),
-        language=realm_user_default.default_language,
+        language=language,
         context=context,
         realm=referrer.realm,
     )
@@ -6791,6 +6790,7 @@ def do_invite_users(
         increment=len(validated_emails),
     )
 
+    realm_user_default = RealmUserDefault.objects.get(realm=user_profile.realm)
     # Now that we are past all the possible errors, we actually create
     # the PreregistrationUser objects and trigger the email invitations.
     for email in validated_emails:
@@ -6802,7 +6802,7 @@ def do_invite_users(
         stream_ids = [stream.id for stream in streams]
         prereg_user.streams.set(stream_ids)
 
-        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id}
+        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id, "email_language": realm_user_default.default_language}
         queue_json_publish("invites", event)
 
     if skipped:
@@ -6920,11 +6920,13 @@ def do_resend_user_invite_email(prereg_user: PreregistrationUser) -> int:
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index c4e66e66d2..b01177e373 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -6624,7 +6624,7 @@ def filter_presence_idle_user_ids(user_ids: Set[int]) -> List[int]:
     return sorted(idle_user_ids)
 
 
-def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile) -> str:
+def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfile, language: str) -> str:
     """
     Send the confirmation/welcome e-mail to an invited user.
     """
@@ -6635,12 +6635,11 @@ def do_send_confirmation_email(invitee: PreregistrationUser, referrer: UserProfi
         "activate_url": activation_url,
         "referrer_realm_name": referrer.realm.name,
     }
-    realm_user_default = RealmUserDefault.objects.get(realm=referrer.realm)
     send_email(
         "zerver/emails/invitation",
         to_emails=[invitee.email],
         from_address=FromAddress.tokenized_no_reply_address(),
-        language=realm_user_default.default_language,
+        language=language,
         context=context,
         realm=referrer.realm,
     )
@@ -6791,6 +6790,7 @@ def do_invite_users(
         increment=len(validated_emails),
     )
 
+    realm_user_default = RealmUserDefault.objects.get(realm=user_profile.realm)
     # Now that we are past all the possible errors, we actually create
     # the PreregistrationUser objects and trigger the email invitations.
     for email in validated_emails:
@@ -6802,7 +6802,7 @@ def do_invite_users(
         stream_ids = [stream.id for stream in streams]
         prereg_user.streams.set(stream_ids)
 
-        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id}
+        event = {"prereg_id": prereg_user.id, "referrer_id": user_profile.id, "email_language": realm_user_default.default_language}
         queue_json_publish("invites", event)
 
     if skipped:
@@ -6920,11 +6920,13 @@ def do_resend_user_invite_email(prereg_user: PreregistrationUser) -> int:
     )
 
     clear_scheduled_invitation_emails(prereg_user.email)
+    realm_user_default = RealmUserDefault.objects.get(realm=prereg_user.realm)
     # We don't store the custom email body, so just set it to None
     event = {
         "prereg_id": prereg_user.id,
         "referrer_id": prereg_user.referred_by.id,
         "email_body": None,
+        "email_language": realm_user_default.default_language,
     }
     queue_json_publish("invites", event)
 
diff --git a/zerver/tests/test_signup.py b/zerver/tests/test_signup.py
index c8cb3fa2f1..a95cd8b2f3 100644
--- a/zerver/tests/test_signup.py
+++ b/zerver/tests/test_signup.py
@@ -1067,7 +1067,7 @@ class InviteUserTest(InviteUserBase):
         #       the large number of queries), so I just
         #       use an approximate equality check.
         actual_count = len(queries)
-        expected_count = 313
+        expected_count = 253
         if abs(actual_count - expected_count) > 1:
             raise AssertionError(
                 f"""
diff --git a/zerver/worker/queue_processors.py b/zerver/worker/queue_processors.py
index d1e3230735..3d9b7ebdd8 100644
--- a/zerver/worker/queue_processors.py
+++ b/zerver/worker/queue_processors.py
@@ -425,7 +425,7 @@ class ConfirmationEmailWorker(QueueProcessingWorker):
         logger.info(
             "Sending invitation for realm %s to %s", referrer.realm.string_id, invitee.email
         )
-        activate_url = do_send_confirmation_email(invitee, referrer)
+        activate_url = do_send_confirmation_email(invitee, referrer, data["email_language"])
 
         # queue invitation reminder
         if settings.INVITATION_LINK_VALIDITY_DAYS >= 4:
@@ -436,13 +436,12 @@ class ConfirmationEmailWorker(QueueProcessingWorker):
                 referrer_email=referrer.delivery_email,
                 referrer_realm_name=referrer.realm.name,
             )
-            realm_user_default = RealmUserDefault.objects.get(realm=referrer.realm)
             send_future_email(
                 "zerver/emails/invitation_reminder",
                 referrer.realm,
                 to_emails=[invitee.email],
                 from_address=FromAddress.tokenized_no_reply_placeholder,
-                language=realm_user_default.default_language,
+                language=data["email_language"],
                 context=context,
                 delay=datetime.timedelta(days=settings.INVITATION_LINK_VALIDITY_DAYS - 2),
             )
```

Applying this diff on the current PR state, leads to increase of 2 queries (i.e. 253).

Reply:
Yeah, that looks great.  Maybe we can do that as a prep commit in a separate PR?  We'll need some code to handle old events already in the queue.

Reply:
Opened #18950.